homeassistant/components/motionmount/config_flow.py

```diff
@@ -0,0 +1,172 @@
+            if self._async_current_entries():
+                return self.async_abort(reason="already_configured")
+
+            # Otherwise we try to continue with the generic uid
+            info[CONF_UUID] = config_entries.DEFAULT_DISCOVERY_UNIQUE_ID
+
+        # If the device mac is valid we use it, otherwise we use the default id
+        if info.get(CONF_UUID, EMPTY_MAC) != EMPTY_MAC:
+            unique_id = info[CONF_UUID]
+        else:
+            unique_id = config_entries.DEFAULT_DISCOVERY_UNIQUE_ID
+
+        name = info.get(CONF_NAME, user_input[CONF_HOST])
+
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured(
+            updates={
+                CONF_HOST: user_input[CONF_HOST],
+                CONF_PORT: user_input[CONF_PORT],
+            }
+        )
+
+        return self.async_create_entry(title=name, data=user_input)
+
+    async def async_step_zeroconf(
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> FlowResult:
+        """Handle zeroconf discovery."""
+
+        # Extract information from discovery
+        host = discovery_info.hostname
+        port = discovery_info.port
+        zctype = discovery_info.type
+        name = discovery_info.name.replace(f".{zctype}", "")
+        unique_id = discovery_info.properties.get("mac")
+
+        self.discovery_info.update(
+            {
+                CONF_HOST: host,
+                CONF_PORT: port,
+                CONF_NAME: name,
+            }
+        )
+
+        if unique_id:
+            # If we already have the unique id, try to set it now
+            # so we can avoid probing the device if its already
+            # configured or ignored
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured(
+                updates={CONF_HOST: host, CONF_PORT: port}
+            )
+
+        self.context.update({"title_placeholders": {"name": name}})
+
+        try:
+            info = await self._validate_input(self.discovery_info)
+        except ConnectionError:
+            return self.async_abort(reason="cannot_connect")
+        except TimeoutError:
+            return self.async_abort(reason="time_out")
+        except motionmount.NotConnectedError:
+            return self.async_abort(reason="not_connected")
+        except motionmount.MotionMountResponseError:
+            info = {}
+            # We continue as we want to be able to connect with older FW that does not support MAC address
+
+        # If the device supplied as with a valid MAC we use that
+        if info.get(CONF_UUID, EMPTY_MAC) != EMPTY_MAC:
+            unique_id = info[CONF_UUID]
+
+        if unique_id:
+            await self.async_set_unique_id(unique_id)
+            self._abort_if_unique_id_configured(
+                updates={CONF_HOST: host, CONF_PORT: port}
+            )
+
+        await self._async_handle_discovery_without_unique_id()
```

Comment:
We should abort here if we don't have a unique id instead.

Reply:
Hi @bdraco, I'm not sure I fully understand why you want to abort. Most of this config flow code is based on the ipp integration, which basically does the same thing. As fas as I understand the contents of the `_async_handle_discovery_without_unique_id()` method, it assigns DEFAULT_DISCOVERY_UNIQUE_ID as the unique id and it makes sure that only a single entity can have that id. It seems that this use case is the exact purpose of this method. Am I mistaken here? Is there something fundamentally different in how the ipp integration handles devices without a unique id?

Reply:
That's a flaw in ipp that we should not copy here

Reply:
Is there some way we can help those users that have older firmware and only a single device? (At the time of writing, that is everybody) If we simply abort in the zeroconf step, the user has no indication that HA has found his device and might be able to control it. The average user will have a hard time figuring out the IP address of the device for manual entry, which is the whole purpose of zeroconf.

From the documentation [here](https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-ids):
_Alternatively, if an integration can't get a unique ID all the time (e.g., multiple devices, some have one, some don't), a helper is available that still allows for discovery, as long as there aren't any instances of the integrations configured yet._

I get the feeling that this is exactly the situation that this call tries to fix. The situation that's mention is at least exactly the situation we're in here.

Reply:
Thanks for pointing that out.  The documentation seems incorrect. That explains why it's being used that way.  I will do some more investigating

Reply:
It looks like this was added in https://github.com/home-assistant/core/pull/36754 to address this exact problem.  The unfortunate problem is that it only works once. If the user has multiple devices without a unique id they won't be able to add the second one via discovery.  I guess thats better than no discovery, but it likely leads to users opening issues when they can't add a second device via discovery.

It also means we can never update the ip address if it changes, or correctly tell two entries apart, but thats a problem we have to live with either way.

Sorry for the side trip here, this is good to proceed assuming the limitations mentioned above are something you can live with.